server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -745,10 +755,14 @@ public void moveShards() {
                     if (logger.isTraceEnabled()) {
                         logger.trace("Moved shard [{}] to node [{}]", shardRouting, targetNode.getRoutingNode());
                     }
+                    if (allocation.hasTargetRelocatingShards()) {
+                        return true;
+                    }
                 } else if (moveDecision.isDecisionTaken() && moveDecision.canRemain() == false) {
                     logger.trace("[{}][{}] can't move", shardRouting.index(), shardRouting.id());
                 }
             }
+            return allocation.hasTargetRelocatingShards();
```

Comment:
We should have returned before this if there was a relocation. Should this be `return false`?

Reply:
Again, I guess not if the relocations were configured > 1.